### PR TITLE
Duration for effects and a test suite for basic effects

### LINKED ISSUE #493: Feature Request: API to specify effect with total duration
## Description

Currently [Effects](https://github.com/flame-engine/flame/blob/master/doc/effects.md) takes speed as argument. There is a use case where user of library want to specify duration instead. 

Currently user has to look into implementation of Effect and reverse the math of calculating `travelTime`. It would be great if library can provide utility function so I can do like:

```
MoveEffect.withDuration(1000, destination, { curve: Curves.easeIn })
```

Another use case is that I want to make `CombinedEffect` where I move and scale the object. I want to start and stop both animation at the same time. 

## Development environment

Flame version: 0.27.0

Flutter doctor:

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 1.22.1, on Mac OS X 10.15.6 19G2021, locale en-US)

[!] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses
[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
[✓] Android Studio (version 4.0)
[✓] VS Code (version 1.48.2)
[✓] Connected device (1 available)

! Doctor found issues in 1 category.
```

### PR DESCRIPTION
**Pull request by spydon**
# Description

This adds duration to the effects, which is how long it should take for an effect to finish (or how long it should take for it to go one iteration if it is infinite).
It also (finally) adds proper testing for the most fundamental effects, `MoveEffect`, `ScaleEffect` and `RotateEffect`. A follow-up PR will come with tests for `SequenceEffect` and `CombinedEffect`.

Fixes #493

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

If something is unclear, please submit the PR anyways and ask about what you thought was unclear.

- [x] This branch is based on `v1.0.0`
- [x] This PR is targeted to merge into `v1.0.0` (not `master`)
- [x] I have added an entry under `[next]` in `CHANGELOG.md`
- [x] I have formatted my code with `flutter format`
- [x] I have made corresponding changes to the documentation
- [x] I have added examples for new features in `doc/examples`
- [x] The continuous integration (CI) is passing

### CODE REVIEW

File: doc/effects.md [hunk 1]
```diff
@@ -20,10 +20,23 @@ If `isInfinite` is false and `isAlternating` is true the effect will go from the
...
+    - pixels/s for `ScaleEffect`
+    - radians/s for `RotateEffect`
+
+One of these two needs to be defined, if both are defined `duration` takes precedence.
+
 ## MoveEffect
 
 Applied to `PositionComponent`s, this effect can be used to move the component to new positions, using an [animation curve](https://api.flutter.dev/flutter/animation/Curves-class.html).
 
+The speed is measured in pixels/s, and remember that you can give `duration` as an argument instead of `speed`.
```
Comment (erickzanardo) [inline doc/effects.md:38]: Nit, in this sentence I would write pixels per second

File: lib/effects/move_effect.dart [hunk 2]
```diff
@@ -18,25 +18,28 @@ class Vector2Percentage {
...
-    this.curve,
+    double duration,
+    double speed,
+    Curve curve,
     bool isInfinite = false,
     bool isAlternating = false,
     bool isRelative = false,
     void Function() onComplete,
-  }) : super(
+  })  : assert(duration != null || speed != null),
```
Comment (erickzanardo) [inline lib/effects/move_effect.dart:36]: Maybe a message here?

File: lib/effects/rotate_effect.dart [hunk 3]
```diff
@@ -3,24 +3,27 @@ import 'package:meta/meta.dart';
...
+    @required this.angle, // As many radians as you want to rotate
+    double duration, // How long it should take for completion
+    double speed, // The speed of rotation in radians/s
+    Curve curve,
     bool isInfinite = false,
     bool isAlternating = false,
     bool isRelative = false,
     void Function() onComplete,
-  }) : super(
+  })  : assert(duration != null || speed != null),
```
Comment (erickzanardo) [inline lib/effects/rotate_effect.dart:20]: Maybe a message here?

File: lib/effects/scale_effect.dart [hunk 4]
```diff
@@ -4,24 +4,27 @@ import 'package:meta/meta.dart';
...
-    this.curve,
+    double duration, // How long it should take for completion
+    double speed, // The speed of the scaling in pixels/s
+    Curve curve,
     bool isInfinite = false,
     bool isAlternating = false,
     bool isRelative = false,
     void Function() onComplete,
-  }) : super(
+  })  : assert(duration != null || speed != null),
```
Comment (erickzanardo) [inline lib/effects/scale_effect.dart:21]: Maybe a message here?

**Review by erickzanardo — APPROVED** [review]
Some small optional comments, but LGTM

**Comment by spydon** [comment]
Seems to be drifting in the isInfinite effect still, will have a look later.

File: doc/effects.md [hunk 5]
```diff
@@ -20,10 +20,23 @@ If `isInfinite` is false and `isAlternating` is true the effect will go from the
...
+## Common for MoveEffect, ScaleEffect and RotateEffect (SimplePositionComponentEffects)
+A common thing for `MoveEffect`, `ScaleEffect` and `RotateEffect` is that it takes `duration` and `speed` as arguments.
+
+- Duration means the time it takes for one iteration from beginning to end, with alternation taken into account (but not `isInfinite`).
+- Speed is the speed of the effect
+    - pixels per second for `MoveEffect`
+    - pixels per second for `ScaleEffect`
+    - radians per second for `RotateEffect`
+
+One of these two needs to be defined, if both are defined `duration` takes precedence.
```
Comment (luanpotter) [inline doc/effects.md:32]: should it thrown an error/assert?

File: doc/effects.md [hunk 6]
```diff
@@ -20,10 +20,23 @@ If `isInfinite` is false and `isAlternating` is true the effect will go from the
 
 When an effect is completed the callback `onComplete` will be called, it can be set as an optional argument to your effect.
 
+## Common for MoveEffect, ScaleEffect and RotateEffect (SimplePositionComponentEffects)
+A common thing for `MoveEffect`, `ScaleEffect` and `RotateEffect` is that it takes `duration` and `speed` as arguments.
+
+- Duration means the time it takes for one iteration from beginning to end, with alternation taken into account (but not `isInfinite`).
```
Comment (luanpotter) [inline doc/effects.md:26]: can you use duration if `isInfinite`? maybe you can rephrase to make it clear that if isInfinite is set, you can only use speed.

File: doc/examples/effects/combined_effects/lib/main.dart [hunk 7]
```diff
@@ -35,13 +35,19 @@ class MyGame extends BaseGame with TapDetector {
     final dx = details.localPosition.dx;
     final dy = details.localPosition.dy;
     greenSquare.clearEffects();
+    final Vector2 currentTap = Vector2(dx, dy);
```
Comment (luanpotter) [inline doc/examples/effects/combined_effects/lib/main.dart:38]: nit: `details.localPosition.toVector2()` :p 
I am addicted haha

File: lib/effects/effects.dart [hunk 8]
```diff
@@ -133,15 +142,47 @@ abstract class PositionComponentEffect
...
+  SimplePositionComponentEffect(
+    bool initialIsInfinite,
+    bool initialIsAlternating, {
+    this.duration,
+    this.speed,
+    Curve curve,
+    bool isRelative = false,
+    void Function() onComplete,
+  })  : assert(
+          duration != null || speed != null,
```
Comment (luanpotter) [inline lib/effects/effects.dart:178]: this could be a xor

File: lib/effects/scale_effect.dart [hunk 9]
```diff
@@ -4,43 +4,52 @@ import 'package:meta/meta.dart';
...
-  Curve curve;
   Vector2 _startSize;
   Vector2 _delta;
 
   ScaleEffect({
     @required this.size,
-    @required this.speed,
-    this.curve,
+    double duration, // How long it should take for completion
+    double speed, // The speed of the scaling in pixels per second
```
Comment (luanpotter) [inline lib/effects/scale_effect.dart:15]: I would add to these comments the fact that it's one xor the other (not only to the docs)

File: test/effects/effect_test_utils.dart [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    const double floatRange = 0.001;
+    bool acceptableVector(Vector2 vector, Vector2 expectedVector) {
+      return (expectedVector - vector).length < floatRange;
+    }
+
+    final bool acceptablePosition =
+        acceptableVector(component.position, expectedPosition);
+    final bool acceptableSize = acceptableVector(component.size, expectedSize);
+    final bool acceptableAngle =
+        (expectedAngle - component.angle).abs() < floatRange;
```
Comment (luanpotter) [inline test/effects/effect_test_utils.dart:54]: I wonder if dart test thing has a method to do this. most test frameworks have a assertDouble sort of thing with customizable precision

File: test/effects/scale_effect_test.dart [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+import 'effect_test_utils.dart';
+
+void main() {
+  final Random random = Random();
+  Vector2 randomVector2() => (Vector2.random(random) * 100)..round();
+  final Vector2 argumentSize = randomVector2();
+  TestComponent component() => TestComponent(size: randomVector2());
+
+  ScaleEffect effect(bool isInfinite, bool isAlternating) {
```
Comment (luanpotter) [inline test/effects/scale_effect_test.dart:16]: nit: I think making these named parameters will make the callers more readable

File: test/effects/rotate_effect_test.dart [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+import 'package:flame/effects/effects.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+import 'effect_test_utils.dart';
+
+void main() {
+  const double angleArgument = 6.0;
+  TestComponent component() => TestComponent(angle: 0.5);
+
+  RotateEffect effect(bool isInfinite, bool isAlternating) {
```
Comment (luanpotter) [inline test/effects/rotate_effect_test.dart:11]: nit: I think making these named parameters will make the callers more readable

File: test/effects/move_effect_test.dart [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+import 'effect_test_utils.dart';
+
+void main() {
+  final Random random = Random();
+  Vector2 randomVector2() => (Vector2.random(random) * 100)..round();
+  final List<Vector2> path = List.generate(3, (i) => randomVector2());
+  TestComponent component() => TestComponent(position: randomVector2());
+
+  MoveEffect effect(bool isInfinite, bool isAlternating) {
```
Comment (luanpotter) [inline test/effects/move_effect_test.dart:16]: nit: I think making these named parameters will make the callers more readable

File: doc/effects.md (hunk 6 above)
Comment (spydon) [reply doc/effects.md:26]: You can still use it, you set the length of one iteration if the effect is infinite.
That is what I'm trying to say in this sentence, maybe it was not very clearly formulated.

File: doc/examples/effects/combined_effects/lib/main.dart (hunk 7 above)
Comment (spydon) [reply doc/examples/effects/combined_effects/lib/main.dart:38]: Ah nice, forgot about these old test when I added the vector extension.
I'll add them now. :)

File: lib/effects/effects.dart (hunk 8 above)
Comment (spydon) [reply lib/effects/effects.dart:178]: Good idea

File: test/effects/effect_test_utils.dart (hunk 10 above)
Comment (spydon) [reply test/effects/effect_test_utils.dart:54]: That would be great, haven't found anything like it though.

File: test/effects/rotate_effect_test.dart (hunk 12 above)
Comment (spydon) [reply test/effects/rotate_effect_test.dart:11]: Good idea, done!

File: test/effects/effect_test_utils.dart (hunk 10 above)
Comment (luanpotter) [reply test/effects/effect_test_utils.dart:64]: what about `https://pub.dev/documentation/matcher/latest/matcher/closeTo.html`?

File: doc/effects.md (hunk 6 above)
Comment (luanpotter) [reply doc/effects.md:29]: I got it now, lg. Maybe add an example of this particular case? as it is the more non-obvious one

File: test/effects/effect_test_utils.dart (hunk 10 above)
Comment (spydon) [reply test/effects/effect_test_utils.dart]: That worked super, thanks!

File: doc/effects.md (hunk 6 above)
Comment (spydon) [reply doc/effects.md:29]: Added to the docs

File: lib/effects/effects_handler.dart [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+export './move_effect.dart';
+export './rotate_effect.dart';
+export './scale_effect.dart';
+export './sequence_effect.dart';
+
+class EffectsHandler {
+  /// The effects that should run on the component
+  final List<ComponentEffect> _effects = [];
+
+  EffectsHandler();
```
Comment (luanpotter) [inline lib/effects/effects_handler.dart:13]: is this necessary? if there are no other constructors dart generates the default constructor for you I think

**Review by luanpotter — APPROVED** [review]
LGTM

File: lib/effects/effects_handler.dart (hunk 14 above)
Comment (spydon) [reply lib/effects/effects_handler.dart:13]: True, there was something in there at first and I just forgot to remove it. Updated now.

**Pull request merged**